website/blog/2023-02-21.ferretdb-compare-arrays.md

```diff
@@ -0,0 +1,530 @@
+---
+title: "How MongoDB compares arrays in filtering"
+slug: ferretdb-compare-arrays
+authors: [chi]
+description: In this article, we explore in detail how MongoDB BSON arrays are compared
+in filter operators.
+keywords:
+  - MongoDB sorting
+  - MongoDB array comparison
+image: /img/blog/ferretdb-how-mongodb-compares-arrays.jpg
+unlisted: true
+---
+
+In this article, we explore in detail how MongoDB BSON arrays are compared in filter operators.
+
+![How MongoDB compares arrays](/img/blog/ferretdb-how-mongodb-compares-arrays.jpg)
+
+<!--truncate-->
+
+## Understanding BSON Comparison for Filtering
+
+The purpose of comparing BSON values is to determine which value is
+equal, greater or less than the other.
+The comparison is used in various operations, such as sorting, filter operators,
+as well as update operators.
+When comparing values of different BSON types, the [BSON Comparison Order](#bson-comparison-order)
+is used to determine the result.
+However, in the case of array comparison against other BSON value,
+the comparison may use the BSON comparison order or an element from the array.
+
+In this blog post, we will discuss the comparison of BSON values used in filter operators,
+with a focus on arrays.
+We will provide examples to help clarify the process.
+
+### BSON Comparison Order
+
+BSON values have a defined order of comparison, listed from lowest to highest as follows:
+
+1. Null
+2. Numbers
+3. String
+4. Object
+5. Array
+6. BinData
+7. ObjectId
+8. Boolean
+9. Date
+10. Timestamp
+11. Regular Expression
+
+The BSON comparison order is used for comparing different BSON types.
+A BSON array
+may use an element from an array to compare with other BSON type.
+We will discuss in detail in the following section.
+
+All number types are the same BSON comparison order, regardless of
+`double`, `int32`, `int64` or `decimal`.
+
+### Comparison Rules of Filter
```

Comment:
It would be nice to have some ordered list summary after text explanation for every set of operators.
I'm talking about something like:
```md
1. Array items are compared by the BSON comparison order
2. Array items are compared by their values
3. If there are no more elements in one of arrays, the shorter one is smaller
```
I'm not sure if those summary should be in this section, or later. I'm not even sure if they are relevant, I guess we can discuss that!